## Content

## 01. Import libraries
## 02. Import data
## 03. Data Consistency Checks - Part 1
## 04. Data Consistency Checks - Part 2 for order dataframe

# 01. Import libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create path

path = r'C:\Users\Frederick\Documents\07-12-2023 Instacart Basket Analysis'

# 02. Import data

In [3]:
# Import data for products

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Import data for orders

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Data Consistency Checks - Part 1

In [5]:
# Create a dataframe

df_test = pd.DataFrame()

In [7]:
# Create a mixed type column

df_test['mix'] = ['a','b',1,True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\Frederick\AppData\Local\Temp\ipykernel_49136\772068622.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [10]:
# Finding missing values in product dataframe

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Create subset of missing value in product name column

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
# Looking at number of rows in current dataframe

df_prods.shape

(49693, 5)

In [14]:
# Creating subset with product dataframe with no missing value

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
# Looking at number of rows(shape) in new product dataframe

df_prods_clean.shape

(49677, 5)

In [16]:
# Create subset with duplicate rows

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
# Number of rows in df_prods_clean

df_prods_clean.shape

(49677, 5)

In [19]:
# Create subset with no duplicate rows

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
# Number of rows in df_prods_clean with no duplicates

df_prods_clean_no_dups.shape

(49672, 5)

# 04. Data Consistency Checks - Part 2 for order dataframe

In [21]:
df_ords.describe()

Unnamed: 0      order_id       user_id  user_order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05       1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04       1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00       1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04       5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05       1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05       2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05       1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

## 02. From the above table, all the statistics look pretty reasonable apart from the maximum value for user order number. The minimum order number is 1 and 75th percentile is 23. However the maximum order is 100 which seems a bit high. So it might be worthwhile to take a look at the user_order_number column to investigate further. 

In [22]:
# Check for mixed types in order dataframe

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\Frederick\AppData\Local\Temp\ipykernel_49136\725199178.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Frederick\AppData\Local\Temp\ipykernel_49136\725199178.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Frederick\AppData\Local\Temp\ipykernel_49136\725199178.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Frederick\AppData\Local\Temp\ipykernel_49136\725199178.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Frederi

## Q4. There are no mix-types data in the order dataframe, as the check mix types function above didn't return any column(s).

In [23]:
# Finding missing values in order dataframe

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
user_order_number              0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## Q5. The only missing values is in days since prior order column, and there are a lot of them - total of 206209 records! One possible explanation is that some customers only order once so there are no data captured on the first order.

In [24]:
# Create subset of missing value in days_since_prior_order column

df_nan_order = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [27]:
df_nan_order.head(15)

Unnamed: 0  order_id  user_id  user_order_number  orders_day_of_week  \
0             0   2539329        1                  1                   2   
11           11   2168274        2                  1                   2   
26           26   1374495        3                  1                   1   
39           39   3343014        4                  1                   6   
45           45   2717275        5                  1                   3   
50           50   2086598        6                  1                   5   
54           54   2565571        7                  1                   3   
75           75    600894        8                  1                   6   
79           79    280530        9                  1                   1   
83           83   1224907       10                  1                   2   
89           89   2411567       11                  1                   0   
97           97   1483446       12                  1                   1   
103         103   2618231       13                  1                   6   
116         116   2906263       14                  1                   2   
130         130   2514324       15                  1                   5   

     order_hour_of_day  days_since_prior_order  
0                    8                     NaN  
11                  11                     NaN  
26                  14                     NaN  
39                  11                     NaN  
45                  12                     NaN  
50                  18                     NaN  
54                   9                     NaN  
75                   0                     NaN  
79                  17                     NaN  
83                  14                     NaN  
89                  11                     NaN  
97                  12                     NaN  
103                 12                     NaN  
116                 18                     NaN  
130                 13                     NaN

In [28]:
df_nan_order.tail()

Unnamed: 0  order_id  user_id  user_order_number  orders_day_of_week  \
3420930     3420930    969311   206205                  1                   4   
3420934     3420934   3189322   206206                  1                   3   
3421002     3421002   2166133   206207                  1                   6   
3421019     3421019   2227043   206208                  1                   1   
3421069     3421069   3154581   206209                  1                   3   

         order_hour_of_day  days_since_prior_order  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN

## Q6. After creating a subset on the missing values and looking at the top 15 rows and bottom 5 rows, it is likely that all these missing value on days since prior order were for the first order. This is because the user_order_number has a value of 1 for the sample I'm looking, indicating that they are all the first order. In this case, I would just leave the missing value blank as there will be days since prior order in the first order. Another alternative option is I could replace the missing value to 'N/A' and change the data type to string. 

In [29]:
# Create subset with duplicate rows in order table

df_dups_order = df_ords[df_ords.duplicated()]

In [30]:
df_dups_order

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, user_order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## Q7. As can be seen above, there are no duplicate values in the order dataframe (i.e. no rows with exactly the same value in all columns). Therefore, there is no need to address any duplicates data. 

In [31]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

In [32]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))